### PR TITLE
fix: add max, min for different kind but same type of arguments

### LINKED ISSUE #5469: Handle max for same type but different kinds of arguments
```fortran
program max_01
    real(8) :: y
    real :: z
    y = 5.2d0
    z = 9.0
    print *, max(y, z)
end program
```
```console
$ lfortran tests/errors/max_01.f90
warning: Different kinds of args in max0 is a non-standard extension
 --> tests/errors/max_01.f90:6:14
  |
6 |     print *, max(y, z)
  |              ^^^^^^^^^ help: ensure all arguments have the same kind to make it standard


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
LCOMPILERS_ASSERT failed: /home/harshita/Desktop/lfortran/src/libasr/asr_utils.h
function Call_t_body(), line number 5950 at 
ASRUtils::check_equal_type(arg_type, orig_arg_type)
ERROR MESSAGE:
ASRUtils::check_equal_type(r32, r64)
Traceback (most recent call last):
  Binary file "/home/harshita/Desktop/lfortran/src/bin/lfortran", in _start
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2682, in main
    return main_app(argc, argv);
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2624, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 1045, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/harshita/Desktop/lfortran/src/lfortran/fortran_evaluator.cpp", line 382, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10366, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/pass_manager.h", line 403, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/pass_manager.h", line 180, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _passes_db[passes[i]](al, *asr, pass_options);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/intrinsic_function.cpp", line 624, in LCompilers::pass_replace_intrinsic_function(Allocator&, LCompilers::ASR::TranslationUnit_t&, LCompilers::PassOptions const&)
    v.visit_TranslationUnit(unit);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 49583, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    this->visit_symbol(*a.second);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5189, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 4900, in void LCompilers::ASR::visit_symbol_t<LCompilers::ReplaceIntrinsicFunctionsVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ReplaceIntrinsicFunctionsVisitor&)
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 49594, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_Program(LCompilers::ASR::Program_t const&)
    self().transform_stmts(xx.m_body, xx.n_body);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 49576, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::transform_stmts(LCompilers::ASR::stmt_t**&, unsigned long&)
    self().visit_stmt(*m_body[i]);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5206, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 4940, in void LCompilers::ASR::visit_stmt_t<LCompilers::ReplaceIntrinsicFunctionsVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ReplaceIntrinsicFunctionsVisitor&)
    case stmtType::Print: { v.visit_Print((const Print_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 49941, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_Print(LCompilers::ASR::Print_t const&)
    self().call_replacer();
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/intrinsic_function.cpp", line 172, in LCompilers::ReplaceIntrinsicFunctionsVisitor::call_replacer()
    replacer.replace_expr(*current_expr);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 48742, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_StringFormat(down_cast<ASR::StringFormat_t>(x));
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 47573, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_StringFormat(LCompilers::ASR::StringFormat_t*)
    self().replace_expr(x->m_args[i]);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 48534, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicElementalFunction(down_cast<ASR::IntrinsicElementalFunction_t>(x));
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/intrinsic_function.cpp", line 89, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicElementalFunction(LCompilers::ASR::IntrinsicElementalFunction_t*)
    ASR::expr_t* current_expr_ = instantiate_function(al, x->base.base.loc,
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/pass/intrinsic_functions.h", line 6069, in LCompilers::ASRUtils::Max::instantiate_Max(Allocator&, LCompilers::Location const&, LCompilers::SymbolTable*, LCompilers::Vec<LCompilers::ASR::ttype_t*>&, LCompilers::ASR::ttype_t*, LCompilers::Vec<LCompilers::ASR::call_arg_t>&, long)
    body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
  File "/home/harshita/Desktop/lfortran/src/libasr/asr_builder.h", line 834, in LCompilers::ASRUtils::ASRBuilder::Call(LCompilers::ASR::symbol_t*, LCompilers::Vec<LCompilers::ASR::call_arg_t>&, LCompilers::ASR::ttype_t*, LCompilers::ASR::expr_t*)
    return ASRUtils::EXPR(ASRUtils::make_FunctionCall_t_util(al, loc,
  File "/home/harshita/Desktop/lfortran/src/libasr/asr_utils.h", line 6003, in LCompilers::ASRUtils::make_FunctionCall_t_util(Allocator&, LCompilers::Location const&, LCompilers::ASR::symbol_t*, LCompilers::ASR::symbol_t*, LCompilers::ASR::call_arg_t*, unsigned long, LCompilers::ASR::ttype_t*, LCompilers::ASR::expr_t*, LCompilers::ASR::expr_t*, bool)
    Call_t_body(al, a_name, a_args, n_args, a_dt, nullptr, false, nopass);
  File "/home/harshita/Desktop/lfortran/src/libasr/asr_utils.h", line 5950, in LCompilers::ASRUtils::Call_t_body(Allocator&, LCompilers::ASR::symbol_t*, LCompilers::ASR::call_arg_t*, unsigned long, LCompilers::ASR::expr_t*, LCompilers::ASR::stmt_t**, bool, bool)
    LCOMPILERS_ASSERT_MSG( ASRUtils::check_equal_type(arg_type, orig_arg_type),
  File "./stdlib/./stdlib/abort.c", line 79, in abort
  File "./signal/../sysdeps/posix/raise.c", line 26, in raise
  File "./nptl/./nptl/pthread_kill.c", line 43, in pthread_kill
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in <invalid>
Abort: Signal SIGABRT (abort) received

Aborted (core dumped)
$ gfortran tests/errors/max_01.f90 && ./a.out
   9.0000000000000000     
```

### PR DESCRIPTION
**Pull request by HarshilShah1804**
Fix #5469 
As discussed in the issue, the code now gives a warning but executes the max and min correctly.
```fortran
program max_01
    real(8) :: y
    real :: z
    y = 5.2d0
    z = 9.0
    print *, max(y, z)
end program
```
```
(lf) harshil-shah@harshil-shah-HP-Pavilion-Laptop-15-eg2xxx:~/lfortran$ src/bin/lfortran tests/errors/max_01.f90
warning: Different kinds of args in max0 is a non-standard extension
 --> tests/errors/max_01.f90:6:14
  |
6 |     print *, max(y, z)
  |              ^^^^^^^^^ help: ensure all arguments have the same kind to make it standard


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
9.00000000000000000e+00
```
Continuation of https://github.com/lfortran/lfortran/pull/5696

### CODE REVIEW

File: tests/tests.toml [hunk 1]
```diff
@@ -3958,19 +3958,19 @@ filename = "errors/string_slice2.f90"
...
 
 [[test]]
-filename = "errors/min_01.f90"
+filename = "../integration_tests/min_01.f90"
 asr = true
 
 [[test]]
-filename = "errors/min_02.f90"
+filename = "../integration_tests/min_02.f90"
 asr = true
```
Comment (Pranavchiku) [inline tests/tests.toml:3974]: we don't need this now, if these are generating binary and running fine remove those from reference tests.

**Review by Pranavchiku — APPROVED** [review]
left a minor change, looks good to me

File: tests/tests.toml (hunk 1 above)
Comment (HarshitaKalani) [reply tests/tests.toml:3974]: Yes, or maybe you can add these in the tests/warnings folder?

**Review by HarshitaKalani — APPROVED** [review]
lgtm, thanks for the PR!

**Comment by certik** [comment]
Looks good. I think you might need to move these into integration_tests from tests/warnings.

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -5964,6 +5964,46 @@ static inline ASR::asr_t* create_SetRemove(Allocator& al, const Location& loc,
...
+        ASR::ttype_t *arg_type = ASRUtils::expr_type(args.p[i]);
+        if (ASR::is_a<ASR::Real_t>(*arg_type)) {
+            if (ASR::is_a<ASR::RealConstant_t>(*args.p[i])) {
+                args.p[i] = EXPR(ASR::make_RealConstant_t(
+                    al, loc, ASR::down_cast<ASR::RealConstant_t>(args.p[i])->m_r,
+                    ASRUtils::TYPE(ASR::make_Real_t(al, loc, target_kind))));
+            } else {
+                args.p[i] = EXPR(ASR::make_Cast_t(
+                    al, loc, args.p[i], ASR::cast_kindType::RealToReal,
+                    ASRUtils::TYPE(ASR::make_Real_t(al, loc, target_kind)), nullptr));
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:5988]: Here you only do this if `args[1]`'s kind is not equal to `target_kind`.

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -5964,6 +5964,46 @@ static inline ASR::asr_t* create_SetRemove(Allocator& al, const Location& loc,
...
+        int kind = ASRUtils::extract_kind_from_ttype_t(arg_type);
+        if (kind > target_kind) {
+            target_kind = kind;
+        }
+    }
+
+    for (size_t i = 0; i < args.size(); i++) {
+        ASR::ttype_t *arg_type = ASRUtils::expr_type(args.p[i]);
+        if (ASR::is_a<ASR::Real_t>(*arg_type)) {
+            if (ASR::is_a<ASR::RealConstant_t>(*args.p[i])) {
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:5981]:
```suggestion
            if (ASR::is_a<ASR::RealConstant_t>(*args[i])) {
```

File: tests/reference/asr-max_01-415ceb0.stdout [hunk 4]
```diff
@@ -65,8 +65,18 @@
...
                             [(IntrinsicElementalFunction
                                 Max
-                                [(Var 2 y)
-                                (Var 2 z)]
+                                [(Cast
+                                    (Var 2 y)
+                                    RealToReal
+                                    (Real 8)
+                                    ()
+                                )
```
[the remark this answers is not in the document]
Comment (gxyd) [reply tests/reference/asr-max_01-415ceb0.stdout:73]: Ah right, I'm sorry, I missed this in my review as well.

**Comment by certik** [comment]
@HarshilShah1804 the PR looks good, but I think it is always inserting casts, which it should not. Let's fix that, after that we can merge it.

File: src/libasr/pass/intrinsic_functions.h (hunk 2 above)
Comment (HarshilShah1804) [reply src/libasr/pass/intrinsic_functions.h:5992]: Changed

File: src/libasr/pass/intrinsic_functions.h (hunk 3 above)
Comment (HarshilShah1804) [reply src/libasr/pass/intrinsic_functions.h]: Changed.

**Comment by certik** [comment]
Thanks!

**Pull request merged**